### ISSUE TITLE
Spec should clarify how to handle errors

### ISSUE THREAD

**Issue by paulcwarren** [container-storage-interface/spec#63]
As the spec stands at the moment when we generate our grpc-go we end up with two possible error channels that we could use.  Firstly, each grpc interface method has an error return value; i.e. 

```golang
type ControllerServer interface {
	CreateVolume(context.Context, *CreateVolumeRequest) (*CreateVolumeResponse, error)
```

Second, the spec also defines a set of Error structs, one for each operation; i.e.

```golang
type CreateVolumeResponse struct {
	// One of the following fields MUST be specified.
	//
	// Types that are valid to be assigned to Reply:
	//	*CreateVolumeResponse_Result_
	//	*CreateVolumeResponse_Error
	Reply isCreateVolumeResponse_Reply `protobuf_oneof:"reply"`
}
```

Accordingly, we should clarify when to use each error so that it is clear to the Plugin developer (or certification test developer) when to use each error channel.  

Our assumption (at DellEMC) is that the struct Error should be used by the plugin developer and the grpc error should NOT (in the plugin's code).  However, it may be used by the underlying grpc framework when unexpected transport errors occur for example when the socket unexpectedly closes.

by way of an example how should the developer handle missing volume name on the CreateVolume call.  Like this?

```golang
if volName == "" {
	return &CreateVolumeResponse{// error struct here}, nil
}
```

or this?
```golang
if volName == "" {
	return &CreateVolumeResponse{}, errors.New("Missing mandatory 'volume_name'")
}
```

or perhaps even like this?
```golang
if volName == "" {
	return &CreateVolumeResponse{// error struct here}, errors.New("Missing mandatory 'volume_name'")
}
```

**Comment by jdef** [comment]
In it's current form, the intent of the spec is that plugin-level errors are reported in xxxResponse objects, so in other words:
```
return &someResponse{/*error struct here*/}, nil
```

FWIW @saad-ali is working on a refactoring of the error handling strategies defined in the spec

**Comment by akutz** [comment]
Hi @paulcwarren,

FWIW, there *is* some code [in Go](https://github.com/codedellemc/gocsi/blob/master/errors.go#L639-L1193) that will parse the errors correctly. For example:

```go
// CheckResponseErrCreateVolume returns a Go error for the
// error message inside of a CSI response if present; otherwise nil
// is returned.
func CheckResponseErrCreateVolume(
	ctx context.Context,
	method string,
	response *csi.CreateVolumeResponse) error {

	rErr := response.GetError()
	if rErr == nil {
		return nil
	}

	if method == "" {
		method = FMCreateVolume
	}

	if err := rErr.GetCreateVolumeError(); err != nil {
		return &Error{
			FullMethod:  method,
			Code:        int32(err.ErrorCode),
			Description: err.ErrorDescription,
		}
	}

	if err := rErr.GetGeneralError(); err != nil {
		return &Error{
			FullMethod:  method,
			Code:        int32(err.ErrorCode),
			Description: err.ErrorDescription,
		}
	}

	return &Error{
		FullMethod:  method,
		Code:        ErrorNoCode,
		Description: rErr.String(),
	}
}
```

This returns a Go `error` that contains the CSI error code, description, and the full method name.

**Comment by julian-hj** [comment]
I believe that this can be closed now that #115 is merged.
